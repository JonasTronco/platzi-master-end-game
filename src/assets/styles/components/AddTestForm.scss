@import '../main.scss';

.add-test-form {
  width: 100%;
  text-align: center;
  padding: $space_x2;
  @include colCenCen;

  &__title {
    @include flexStaCen;
    margin-bottom: $space_x3;

    &__icon {
      width: $space_x9;
      height: $space_x9; 
      fill: $secondary_color;
    }
  }

  &__user-info {
    @include flexCenCen;
    text-align: left;
    margin-bottom: $space_x3;

    img {
      width: $space_x10;
      height: $space_x10;
      border-radius: $space_x5;
      margin-right: $space_x2;
    }
  }

  &__form {

    .input__label {
      margin-top: $space_x2;
    }
    
    &__container {
      width: 100%;
      
      .btn {
        margin-top: $space_x3;
      }
    }
  }
}

.container__test {
  margin-top: $space_x3;
}

.colum__test {
  width: 50%;
}
