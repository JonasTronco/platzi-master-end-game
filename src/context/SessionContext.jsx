import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
