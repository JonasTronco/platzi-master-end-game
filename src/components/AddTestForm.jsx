import PropTypes from 'prop-types';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { useSession } from '../context/SessionContext';
import useRequest from '../hooks/useRequest'
import FileIcon from '../assets/icons/file.svg';
import '../assets/styles/components/AddTestForm.scss';

function AddTestForm({ imageURL, name, lastName}) {
  const { session } = useSession();
  const { response, loading, error } = useRequest(
    session.token,
    `${process.env.NEXT_PUBLIC_API_URL}/exams/`,
    1,
  );

  return (
    <main className="add-test-form">
      <div className="add-test-form__title">
        <FileIcon className="add-test-form__title__icon" />
        <h2 className="add-test-form__title__text">Assign a new exam</h2>
      </div>
      <div className="add-test-form__user-info">
        <img alt="user whose new test is filling" src={imageURL} />
        <strong>
          {name}
          <br />
          {lastName}
        </strong>
      </div>
      <div className="add-test-form__form__container">
        <Formik
          initialValues={{ exam: '', appointment: '' }}
          validate={(values) => {
            const errors = {};
            if (!values.exam) {
              errors.exam = 'Required: please select an exam';
            }
            if (!values.appointment) {
              errors.appointment = 'Required: please select an appointment date';
            }
            return errors;
          }}
          onSubmit={async (values, { setSubmitting }) => {
            alert(JSON.stringify(values));
            setSubmitting(false);
          }}
        >
          {({ isSubmitting }) => (
            <>
              { !loading && response.data ? (
                <Form className="add-test-form__form lg">
                  <div className="input lg">
                    <label className="input__label lg">
                      Select an exam
                      <Field
                        as="select"
                        className="input__field lg"
                        type="exam"
                        name="exam"
                        disabled={isSubmitting}
                      >
                        {response.data.map((exam) => {
                          console.log(exam.name);
                          return <option>{exam.name}</option>;
                        })}
                      </Field>
                    </label>
                    <ErrorMessage
                      name="exam"
                      component="div"
                      className="input__error"
                    />
                  </div>
                  <div className="input lg">
                    <label className="input__label lg">
                      Appointment date
                      <Field
                        className="input__field lg"
                        type="appointment"
                        name="appointment"
                        disabled={isSubmitting}
                      />
                    </label>
                    <ErrorMessage
                      name="appointment"
                      component="div"
                      className="input__error"
                    />
                  </div>
                  <button className="btn lg" type="submit" disabled={isSubmitting}>
                    Add new test
                  </button>
                </Form>
              ) : (<div className="loader" />)}
            </>
          )}
        </Formik>
      </div>
    </main>
  );
}

AddTestForm.propTypes = {
  imageURL: PropTypes.string,
  name: PropTypes.string,
  lastName: PropTypes.string,
};

AddTestForm.defaultProps = {
  imageURL: 'https://i.imgur.com/oMJFiLX.jpg',
  name: 'Catalina',
  lastName: 'Flores',
};

export default AddTestForm;
