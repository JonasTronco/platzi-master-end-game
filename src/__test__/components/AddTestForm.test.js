import AddTestForm from '../../components/AddTestForm';
import { create } from 'react-test-renderer';
import { SessionProvider } from '../../context/SessionContext';
describe('AddTestForm Component', () => {
  const obj = {
    session: {
      token:
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI1ZWNkYTc2NjYxYjE3ZTUwYzg3NzIxYWIiLCJ1c2VybmFtZSI6Im1heW5lLnNuYXNlbC40MTcxIiwic2NvcGVzIjpbInNpZ25pbjphdXRoIiwic2lnbnVwOmF1dGgiLCJyZWFkOm1vdmllcyIsInJlYWQ6dXNlci1tb3ZpZXMiLCJjcmVhdGU6dXNlci1tb3ZpZXMiLCJkZWxldGU6dXNlci1tb3ZpZXMiXSwiaWF0IjoxNTkwNzI1MTIwLCJleHAiOjE1OTExNTcxMjB9.UjjAPEHWbqyPEwX6qcs2UhLvQ6rAeQIL-ME1LDnWBRo',
      user: {
        id: '5ecda76661b17e50c87721ab',
        username: 'mayne.snasel.4171',
        typeOfUser: 'Patient',
        isActive: true,
        imageURL: 'http://dummyimage.com/200x200.png/ff4444/ffffff',
        firstName: 'Mayne',
        lastName: 'Snasel',
        defaultPath: '/tests'
      }
    }
  };
  test('Should render AddTestForm component', () => {
    const component = create(
      <SessionProvider value={obj}>
        <AddTestForm userID={''} />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
