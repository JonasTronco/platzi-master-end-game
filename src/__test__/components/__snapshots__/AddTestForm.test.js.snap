// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTestForm Component Should render AddTestForm component 1`] = `
<main
  className="add-test-form"
>
  <div
    className="add-test-form__user-info"
  >
    <div
      className="loader"
    />
  </div>
  <div
    className="add-test-form__form__container"
  >
    <div
      className="loader"
    />
  </div>
</main>
`;
